src/coprocessor/endpoint.rs

```diff
@@ -256,62 +308,85 @@ impl BatchRunnable<Task> for Host {
-                    };
-
-                    if self.running_task_count() >= self.max_running_task_count {
-                        notify_batch_failed(Error::Full(self.max_running_task_count), reqs);
-                        continue;
+                    self.handle_snapshot_result(q_id, snap_res);
+                }
+                Task::BatchSnapRes(batch) => {
+                    for (q_id, snap_res) in batch {
+                        self.handle_snapshot_result(q_id, snap_res);
                     }
-
-                    for req in reqs {
-                        let pri = req.priority();
-                        let pri_str = get_req_pri_str(pri);
-                        let type_str = get_req_type_str(req.req.get_tp());
-                        COPR_PENDING_REQS.with_label_values(&[type_str, pri_str]).add(1.0);
-                        let end_point = TiDbEndPoint::new(snap.clone());
-                        let txn_id = req.start_ts.unwrap_or_default();
-
-                        if pri == CommandPri::Low {
-                            self.low_priority_pool.execute(txn_id, move || {
-                                end_point.handle_request(req);
-                                COPR_PENDING_REQS.with_label_values(&[type_str, pri_str]).dec();
-                            });
-                        } else if pri == CommandPri::High {
-                            self.high_priority_pool.execute(txn_id, move || {
-                                end_point.handle_request(req);
-                                COPR_PENDING_REQS.with_label_values(&[type_str, pri_str]).dec();
-                            });
+                }
+                Task::RetryRequests(retry) => {
+                    for id in retry {
+                        let reqs = self.reqs.remove(&id).unwrap();
+                        let sched = self.sched.clone();
+                        if let Err(e) =
+                               self.engine.async_snapshot(reqs[0].req.get_context(),
+                                                          box move |(_, res)| {
+                                                              sched.schedule(Task::SnapRes(id, res))
+                                                                  .unwrap()
+                                                          }) {
+                            notify_batch_failed(e, reqs);
                         } else {
-                            self.pool.execute(txn_id, move || {
-                                end_point.handle_request(req);
-                                COPR_PENDING_REQS.with_label_values(&[type_str, pri_str]).dec();
-                            });
+                            self.reqs.insert(id, reqs);
                         }
                     }
                 }
             }
         }
+
+        if grouped_reqs.is_empty() {
+            return;
+        }
+
+        let mut req_ids = Vec::with_capacity(grouped_reqs.len());
```

Comment:
seem we don't need req_ids here.
We can ensure that the first id is self.last_req_id + 1, and the last is self.last_req_id + len(grouped_reqs)